editor/src/clj/editor/pose.clj

```diff
@@ -0,0 +1,255 @@
+(ns editor.pose
+  (:require [clojure.spec.alpha :as s]
+            [editor.math :as math])
+  (:import [javax.vecmath Matrix4d]))
+
+(set! *warn-on-reflection* true)
+(set! *unchecked-math* :warn-on-boxed)
+
+(s/def ::num number?)
+(s/def ::num3 (s/tuple ::num ::num ::num))
+(s/def ::num4 (s/tuple ::num ::num ::num ::num))
+(s/def ::translation ::num3)
+(s/def ::rotation ::num4)
+(s/def ::scale ::num3)
+(s/def ::euler-rotation ::num3)
+
+(defrecord Pose [translation rotation scale])
+
+(defn pose? [value]
+  (instance? Pose value))
+
+(def default-translation (vector-of :double 0.0 0.0 0.0))
+(def default-rotation (vector-of :double 0.0 0.0 0.0 1.0))
+(def default-scale (vector-of :double 1.0 1.0 1.0))
+(def ^Pose default (->Pose default-translation default-rotation default-scale))
+
+(def ^:private mat4-identity (doto (Matrix4d.) (.setIdentity)))
+(def ^:private num4-zero (vector-of :double 0.0 0.0 0.0 0.0))
+
+(defn- significant-translation? [value]
+  (if-some [[^double x ^double y ^double z] value]
+    (not (and (zero? x)
+              (zero? y)
+              (zero? z)))
+    false))
+
+(defn- significant-rotation? [value]
+  (if-some [[^double x ^double y ^double z ^double w] value]
+    (not (and (zero? x)
+              (zero? y)
+              (zero? z)
+              (== 1.0 w)))
+    false))
+
+(defn- significant-scale? [value]
+  (if-some [[^double x ^double y ^double z] value]
+    (not (and (== 1.0 x)
+              (== 1.0 y)
+              (== 1.0 z)))
+    false))
```

Comment:
I think the `nil` check semantically should belong higher up the stack, e.g. in a place where some pose attribute might be `nil`, we do the check before calling these predicates.